### PR TITLE
OWL Axiom support in stated taxonomy and ECL queries

### PR DESCRIPTION
**Pull request by cmark**
This PR adds support for OWL Axiom based stated form.

* Port stated ECL support from https://github.com/IHTSDO/snow-owl fork.
* Add `snomed-owl-toolkit` dependency (version 2.3.2).
* Add new `classAxiomRelationships` and `gciAxiomRelationships` fields to reference set members where the owlExpression field is populated with valid OWL Axiom and can be converted to relationship fragments.
* Use the newly added fields to query stated taxonomy for `statedParents` and `statedAncestors` updates.
* Use the newly added fields to query stated relationships in stated ECL evaluation.

### CODE REVIEW

File: net4j/com.b2international.snowowl.eventbus/src/com/b2international/snowowl/eventbus/EventBusUtil.java [hunk 1]
```diff
@@ -86,7 +87,16 @@ public void handle(IMessage message) {
...
+		final IEventBus bus = new EventBus(name, numberOfWorkers, numberOfWorkers == 0 ? ExecutorServiceFactory.DIRECT : new WorkerExecutorServiceFactory());
+		LifecycleUtil.activate(bus);
+		return bus;
+	}
+	
+	/**
+	 * @return an {@link EventBus} with the specified description and 1 direct thread worker.
+	 */
+	public static IEventBus getDirectBus(String name) {
+		final IEventBus bus = new EventBus(name, 1, ExecutorServiceFactory.DIRECT);
```
Comment (apeteri) [inline net4j/com.b2international.snowowl.eventbus/src/com/b2international/snowowl/eventbus/EventBusUtil.java:99]: What is the difference between `name, 1, ExecutorServiceFactory.DIRECT` and `name, 0` as arguments?

File: snomed/com.b2international.snowowl.snomed.datastore/src/com/b2international/snowowl/snomed/core/ecl/SnomedEclRefinementEvaluator.java [hunk 2]
```diff
@@ -93,17 +106,20 @@
  */
 final class SnomedEclRefinementEvaluator {
 
-	static final Set<String> ALLOWED_CHARACTERISTIC_TYPES = ImmutableSet.of(Concepts.INFERRED_RELATIONSHIP, Concepts.ADDITIONAL_RELATIONSHIP);
+	static final Set<String> STATED_CHARACTERISTIC_TYPES = ImmutableSet.of(Concepts.STATED_RELATIONSHIP, Concepts.ADDITIONAL_RELATIONSHIP);
+	static final Set<String> INFERRED_CHARACTERISTIC_TYPES = ImmutableSet.of(Concepts.INFERRED_RELATIONSHIP, Concepts.ADDITIONAL_RELATIONSHIP);
```
Comment (apeteri) [inline snomed/com.b2international.snowowl.snomed.datastore/src/com/b2international/snowowl/snomed/core/ecl/SnomedEclRefinementEvaluator.java:110]: Do we want to make the distinction between grouped and ungrouped additional properties in the inferred form?

File: snomed/com.b2international.snowowl.snomed.datastore/src/com/b2international/snowowl/snomed/core/ecl/SnomedEclRefinementEvaluator.java [hunk 3]
```diff
@@ -564,36 +579,121 @@ public SnomedEclRefinementEvaluator(EclExpression focusConcepts) {
...
-				.then(input -> input.stream().map(r -> new Property(r.getId(), r.getSourceId(), r.getTypeId(), r.getDestinationId(), r.getGroup())).collect(Collectors.toSet()));
+		Promise<Collection<Property>> then = req.build(context.id(), context.branchPath())
+			.execute(context.service(IEventBus.class))
+			.then(input -> input.stream().map(r -> new Property(r.getSourceId(), r.getTypeId(), r.getDestinationId(), r.getGroup())).collect(Collectors.toSet()));
+		
+		if (Trees.STATED_FORM.equals(expressionForm)) {
+			final Set<Property> axiomStatements = evalAxiomStatements(context, groupedRelationshipsOnly, sourceIds, typeIds, destinationIds);
+			return then.then(relationshipStatements -> ImmutableSet.<Property>builder().addAll(relationshipStatements).addAll(axiomStatements).build());
+		} else {
+			return then;
```
Comment (apeteri) [inline snomed/com.b2international.snowowl.snomed.datastore/src/com/b2international/snowowl/snomed/core/ecl/SnomedEclRefinementEvaluator.java]: Please find another name for this local variable.

File: snomed/com.b2international.snowowl.snomed.datastore/src/com/b2international/snowowl/snomed/core/ecl/SnomedEclRefinementEvaluator.java [hunk 4]
```diff
@@ -564,36 +579,121 @@ public SnomedEclRefinementEvaluator(EclExpression focusConcepts) {
...
+			
+			ExpressionBuilder activeOwlAxiomMemberQuery = Expressions.builder()
+					.filter(active())
+					.filter(Expressions.nestedMatch(SnomedRefSetMemberIndexEntry.Fields.CLASS_AXIOM_RELATIONSHIP, axiomFilter.build()));
+			
+			if (sourceIds != null) {
+				activeOwlAxiomMemberQuery.filter(SnomedRefSetMemberIndexEntry.Expressions.referencedComponentIds(sourceIds));
+			}
+			
+			final Query<SnomedRefSetMemberIndexEntry> activeAxiomISARelationshipsQuery = Query.select(SnomedRefSetMemberIndexEntry.class)
```
Comment (apeteri) [inline snomed/com.b2international.snowowl.snomed.datastore/src/com/b2international/snowowl/snomed/core/ecl/SnomedEclRefinementEvaluator.java]: ⚠️ This doesn't seem to be an IS A-only axiom query, there is a `typeId` filter above.

File: snomed/com.b2international.snowowl.snomed.datastore/src/com/b2international/snowowl/snomed/datastore/request/SnomedConceptSearchRequestBuilder.java [hunk 5]
```diff
@@ -106,6 +106,18 @@ public final SnomedConceptSearchRequestBuilder filterByEcl(String expression) {
 		return addOption(SnomedConceptSearchRequest.OptionKey.ECL, expression);
 	}
 	
+	/**
+	 * Filter matches in the stated tree by the specified Expression Constraint Language (ECL) expression. 
+	 * The currently supported ECL version is v1.3. See <a href="http://snomed.org/ecl">ECL Specification and Guide</a> or
+	 * <a href="http://www.snomed.org/news-articles/expression-constraint-language">About ECL</a> for more information.
```
Comment (apeteri) [inline snomed/com.b2international.snowowl.snomed.datastore/src/com/b2international/snowowl/snomed/datastore/request/SnomedConceptSearchRequestBuilder.java]: This link is non-reachable ⚡️

File: snomed/com.b2international.snowowl.snomed.datastore/src/com/b2international/snowowl/snomed/datastore/taxonomy/Edges.java [hunk 6]
```diff
@@ -0,0 +1,51 @@
...
+	Edges(long sourceId, long[] destinationIds) {
+		this.sourceId = sourceId;
+		this.destinationIds = destinationIds;
+		// keep the destinationIds sorted, so Edges will be equal even if the destinationIds long array was not in the same order
+		Arrays.sort(this.destinationIds);
+	}
+	
+	@Override
+	public int hashCode() {
+		return Objects.hash(sourceId, destinationIds);
```
Comment (apeteri) [inline snomed/com.b2international.snowowl.snomed.datastore/src/com/b2international/snowowl/snomed/datastore/taxonomy/Edges.java]: `destinationIds` also needs to be hashed by content:

```suggestion
		return Objects.hash(sourceId, Arrays.hashCode(destinationIds));
```

File: snomed/com.b2international.snowowl.snomed.datastore/src/com/b2international/snowowl/snomed/datastore/taxonomy/Taxonomies.java [hunk 7]
```diff
@@ -48,62 +57,141 @@
...
+		activeIsaRelationships = null;
+		
+		if (Concepts.STATED_RELATIONSHIP.equals(characteristicTypeId)) {
+			// search existing axioms defined for the given set of conceptIds
+			ExpressionBuilder activeOwlAxiomMemberQuery = Expressions.builder()
+					.filter(active())
+					.filter(Expressions.nestedMatch(SnomedRefSetMemberIndexEntry.Fields.CLASS_AXIOM_RELATIONSHIP, 
+						Expressions.builder()
+							.filter(typeId(Concepts.IS_A))
+							.filter(destinationIds(concepts))
```
Comment (apeteri) [inline snomed/com.b2international.snowowl.snomed.datastore/src/com/b2international/snowowl/snomed/datastore/taxonomy/Taxonomies.java]: ⚠️ For stated IS A relationships above, the destination is not restricted unless `filterByConceptIds` is enabled.

File: snomed/com.b2international.snowowl.snomed.datastore/src/com/b2international/snowowl/snomed/datastore/taxonomy/TaxonomyGraphUpdater.java [hunk 8]
```diff
@@ -211,97 +218,55 @@ public void run() {
...
+	private void updateEdge(SnomedOWLExpressionRefSetMember owlMember, TaxonomyGraph graphToUpdate) {
+		if (owlMember.isActive()) {
+			SnomedOWLExpressionConverterResult result = expressionConverter.toSnomedOWLRelationships(owlMember.getReferencedComponentId(), owlMember.getOwlExpression());
+			if (!CompareUtils.isEmpty(result.getClassAxiomRelationships())) {
+				final long[] destinationIds = result.getClassAxiomRelationships().stream()
+					.filter(r -> Concepts.IS_A.equals(r.getTypeId()))
+					.map(SnomedOWLRelationshipDocument::getDestinationId)
+					.mapToLong(Long::parseLong)
+					.toArray();
+				graphToUpdate.addEdge(owlMember.getUuid(), Long.parseLong(owlMember.getReferencedComponentId()), destinationIds);
```
Comment (apeteri) [inline snomed/com.b2international.snowowl.snomed.datastore/src/com/b2international/snowowl/snomed/datastore/taxonomy/TaxonomyGraphUpdater.java:243]: Are OWL expression destination changes also handled in `addEdge`?

**Review by apeteri — CHANGES_REQUESTED** [review]

File: snomed/com.b2international.snowowl.snomed.datastore/src/com/b2international/snowowl/snomed/core/ecl/SnomedEclRefinementEvaluator.java (hunk 4 above)
Comment (cmark) [reply snomed/com.b2international.snowowl.snomed.datastore/src/com/b2international/snowowl/snomed/core/ecl/SnomedEclRefinementEvaluator.java]: Renamed!

File: snomed/com.b2international.snowowl.snomed.datastore/src/com/b2international/snowowl/snomed/datastore/request/SnomedConceptSearchRequestBuilder.java (hunk 5 above)
Comment (cmark) [reply snomed/com.b2international.snowowl.snomed.datastore/src/com/b2international/snowowl/snomed/datastore/request/SnomedConceptSearchRequestBuilder.java]: Fixed!

File: snomed/com.b2international.snowowl.snomed.datastore/src/com/b2international/snowowl/snomed/datastore/taxonomy/Edges.java (hunk 6 above)
Comment (cmark) [reply snomed/com.b2international.snowowl.snomed.datastore/src/com/b2international/snowowl/snomed/datastore/taxonomy/Edges.java]: Fixed!

File: snomed/com.b2international.snowowl.snomed.datastore/src/com/b2international/snowowl/snomed/datastore/taxonomy/Taxonomies.java (hunk 7 above)
Comment (cmark) [reply snomed/com.b2international.snowowl.snomed.datastore/src/com/b2international/snowowl/snomed/datastore/taxonomy/Taxonomies.java]: Right, excluded `destinationIds` filter in that case.

File: snomed/com.b2international.snowowl.snomed.datastore/src/com/b2international/snowowl/snomed/datastore/taxonomy/TaxonomyGraphUpdater.java (hunk 8 above)
Comment (cmark) [reply snomed/com.b2international.snowowl.snomed.datastore/src/com/b2international/snowowl/snomed/datastore/taxonomy/TaxonomyGraphUpdater.java:243]: Yes!

**Review by apeteri — APPROVED** [review]

**Pull request merged**